src/main/scala/util/DelayQueue.scala

```diff
@@ -0,0 +1,49 @@
+// See LICENSE.SiFive for license details.
+
+package freechips.rocketchip.util
+
+import chisel3._
+import chisel3.util._
+
+object DelayQueue {
+  private def apply[T <: Data](source: DecoupledIO[T], timer: UInt, delay: UInt, depth: Int): DecoupledIO[T] = {
+    val q = Module(new Queue(new Bundle {
```

Comment:
Can you wrap everything in this block in another module? This way we can avoid instances of `q` in the module hierarchy.

Reply:
Ok I did so. Note: I kept the timer outside of the module so that power-users could re-use the timer across different `DelayQueues` (like what #3512 does).